process/block/preprocess/sovereignRewardTxPreProcessor.go

```diff
@@ -0,0 +1,103 @@
+type ArgsRewardTxPreProcessor struct {
+	RewardTxDataPool           dataRetriever.ShardedDataCacherNotifier
+	Store                      dataRetriever.StorageService
+	Hasher                     hashing.Hasher
+	Marshalizer                marshal.Marshalizer
+	RewardProcessor            process.RewardTransactionProcessor
+	ShardCoordinator           sharding.Coordinator
+	Accounts                   state.AccountsAdapter
+	OnRequestRewardTransaction func(shardID uint32, txHashes [][]byte)
+	GasHandler                 process.GasHandler
+	PubkeyConverter            core.PubkeyConverter
+	BlockSizeComputation       BlockSizeComputationHandler
+	BalanceComputation         BalanceComputationHandler
+	ProcessedMiniBlocksTracker process.ProcessedMiniBlocksTracker
+	TxExecutionOrderHandler    common.TxExecutionOrderHandler
+}
+
+// NewSovereignRewardsTxPreProcessor creates a sovereign rewards tx pre-processor
+func NewSovereignRewardsTxPreProcessor(args ArgsRewardTxPreProcessor) (*sovereignRewardsTxPreProcessor, error) {
+	rtp, err := baseCreateRewardTxPreProc(args)
+	if err != nil {
+		return nil, err
+	}
+
+	srtp := &sovereignRewardsTxPreProcessor{
+		rtp,
+	}
+	srtp.rewardTxPool.RegisterOnAdded(srtp.receivedRewardTransaction)
+	return srtp, err
+}
+
+// receivedRewardTransaction is a callback function called when a new reward transaction
+// is added in the reward transactions pool. We do not need to wait for broadcasted txs on a channel, since
+// each node is processing these rewards. Therefore, whenever adding a reward tx in pool, we will directly execute it.
+func (srtp *sovereignRewardsTxPreProcessor) receivedRewardTransaction(key []byte, value interface{}) {
+	tx, ok := value.(data.TransactionHandler)
+	if !ok {
+		log.Warn("sovereignRewardsTxPreProcessor.receivedRewardTransaction", "error", process.ErrWrongTypeAssertion)
+		return
+	}
+
+	err := srtp.receivedRewardTx(key, tx, &srtp.rewardTxsForBlock)
+	if err != nil {
+		log.Error("sovereignRewardsTxPreProcessor.receivedRewardTx", "error", err)
+	}
+}
+
+func (srtp *sovereignRewardsTxPreProcessor) receivedRewardTx(
+	txHash []byte,
+	tx data.TransactionHandler,
+	forBlock *txsForBlock,
+) error {
+	if check.IfNil(tx) {
+		log.Warn("nil rewardsTransaction in sovereignRewardsTxPreProcessor.receivedRewardTx", "hash", txHash)
+		return process.ErrMissingTransaction
+	}
+
+	rwdTx, castOk := tx.(*rewardTx.RewardTx)
+	if !castOk {
+		log.Warn("invalid rewardsTransaction in sovereignRewardsTxPreProcessor.receivedRewardTx", "hash", txHash)
+		return process.ErrWrongTypeAssertion
+	}
+
+	forBlock.mutTxsForBlock.Lock()
+	forBlock.txHashAndInfo[string(txHash)] = &txInfo{
+		tx: tx,
+	}
+	forBlock.mutTxsForBlock.Unlock()
+
+	err := srtp.saveAccountBalanceForAddress(tx.GetRcvAddr())
+	if err != nil {
+		return err
+	}
+
+	srtp.txExecutionOrderHandler.Add(txHash)
+	return srtp.rewardsProcessor.ProcessRewardTransaction(rwdTx)
```

Comment:
This is wrong. This is not going to be executed sequentially and it will create problems.

Reply:
On the SovereignEpochProcessor we can do the addBalance stuff as well.

Reply:
Fixed as suggested